contracts/xasset-orchestrator/src/orchestrator.rs

```diff
@@ -0,0 +1,168 @@
+use crate::{error::Error, Contract};
+use loam_sdk::{
+    loamstorage,
+    soroban_sdk::{
+        self, env, xdr::ToXdr, Address, Bytes, BytesN, Env, InstanceItem, Lazy, LoamKey,
+        PersistentMap, String, Symbol,
+    },
+    subcontract,
+};
+use loam_subcontract_core::Core;
+
+#[loamstorage]
+pub struct Storage {
+    /// XLM SAC contract address; initialized and then passed
+    /// to deployed xasset contracts
+    xlm_sac: InstanceItem<Address>,
+    /// XLM oracle contract, initialized and then passed
+    /// to deployed xasset contracts
+    xlm_contract: InstanceItem<Address>,
+    /// A map of deployed asset contracts to their asset symbol.
+    /// This is used to check if a contract is a valid asset contract
+    /// and to get the asset symbol from the contract address.
+    /// The key is the asset symbol, the value is the asset contract address.
+    assets: PersistentMap<String, Address>,
+}
+
+#[subcontract]
+pub trait IsOrchestratorTrait {
+    fn init(&mut self, xlm_sac: Address, xlm_contract: Address) -> Result<(), Error>;
+    fn deploy_asset_contract(
+        &mut self,
+        asset_contract: Address,
+        pegged_asset: Symbol,
+        min_collat_ratio: u32,
+        name: String,
+        symbol: String,
+        decimals: u32,
+        annual_interest_rate: u32,
+    ) -> Result<Address, Error>;
+    fn get_asset_contract(&self, asset_symbol: String) -> Result<Address, Error>;
+    // Manually set an asset symbol to an existing contract address
+    fn set_asset_contract(
+        &mut self,
+        asset_symbol: String,
+        asset_contract: Address,
+    ) -> Result<(), Error>;
+    // Manually set an existing asset symbol to an existing contract address. Dangerous!
+    // This should only be used when needing to update an existing symbol's contract.
+    fn set_existing_asset_contract(
+        &mut self,
+        asset_symbol: String,
+        asset_contract: Address,
+    ) -> Result<(), Error>;
+}
+
+impl IsOrchestratorTrait for Storage {
+    fn init(&mut self, xlm_sac: Address, xlm_contract: Address) -> Result<(), Error> {
+        Contract::require_auth();
+        self.xlm_sac.set(&xlm_sac);
+        self.xlm_contract.set(&xlm_contract);
+        Ok(())
+    }
+
+    fn deploy_asset_contract(
+        &mut self,
+        asset_contract: Address,
+        pegged_asset: Symbol,
+        min_collat_ratio: u32,
+        name: String,
+        symbol: String,
+        decimals: u32,
+        annual_interest_rate: u32,
+    ) -> Result<Address, Error> {
+        let env = env();
+        Contract::require_auth();
+        // Check if the asset contract is already deployed
+        if self.assets.has(symbol.clone()) {
+            return Err(Error::AssetAlreadyDeployed);
+        }
+
+        // Upload the xasset wasm to prep for deploy
+        let wasm_hash = env.deployer().upload_contract_wasm(xasset::WASM);
```

Comment:
This works well enough for now, but this means that the contract's binary stores the bytes of the other contract in the data section of the Wasm of this contract. Meaning it costs extra every time the contract is invoked as more bytes need to be read from storage.

Reply:
Is there a good alternative?

Reply:
The simplest solution currently would be to upload the contract via the CLI. Then you can pass the returned hash when calling the method. The best solution would be to use the registry to handle the publishing of the Wasm and deploying the contract. Honestly the current solution works well for now because we aren't worried about the cost yet and if we use `import_contract` and have xasset as a dep, then every time we build this contract then we'll get the newest version of the other one.

I'm not sure the current workflow when updating the contract but this should be easiest for now.

Reply:
Alright! The orchestrator itself gets deployed from the CLI/as part of infrastructure deployment, so for option A, it wouldn't be ridiculous to say the orchestrator stores an instance value of the wasm_hash at init, like it does for the xlm_sac and xlm oracle contract …

Would rather not build in silly cost situations

Reply:
https://github.com/EquitXCompany/equitx-project/pull/52/files#diff-ac386b2c9b35b08ef66803567f13e89a95a863607f68ee1c6516d3093015305c

The future deployment flow for bootstrapping testnet